File: advent-of-code/2023/day-01/part1.py
#!/opt/homebrew/bin/python3

'''
The newly-improved calibration document consists of lines of text; 
each line originally contained a specific calibration value that the 
Elves now need to recover. On each line, the calibration value can be 
found by combining the first digit and the last digit (in that order) 
to form a single two-digit number.

For example:

1abc2
pqr3stu8vwx
a1b2c3d4e5f
treb7uchet
In this example, the calibration values of these four lines are 12, 38, 15, and 77. 
Adding these together produces 142.

Consider your entire calibration document. What is the sum of all of 
the calibration values?

approach: use 2 pointers to find the first digit encounter for each line
'''

import argparse

def solve(lines):
    total = 0
    for l in lines:
        start_digit, end_digit = None, None

        # get start digit
        for left in range(len(l)):
            if l[left].isdigit():
                start_digit = l[left]
                break
        if not start_digit:
            raise Exception(f"there was not anything digit in line {l}")
        for right in range(len(l) - 1, -1, -1):
            if l[right].isdigit():
                end_digit = l[right]
                break
        total += int(start_digit + end_digit)
    return total


def main():
    parser = argparse.ArgumentParser(description='Advent of code day 1 part 1')
    parser.add_argument('-i', '--input', required=True)
    args = parser.parse_args()

    try:
        with open(args.input, 'r') as input_file:
            lines = [line.strip() for line in input_file]
            print(solve(lines))
    except FileNotFoundError:
        print(f"Input file not found: {args.input}")
    except Exception as e:
        print(f"Encounter error for input: {args.input}, err: {e}")

if __name__ == "__main__":
    main()

File: advent-of-code/2023/day-01/part1.go
package main

import (
  "bufio"
  "flag"
  "fmt"
  "os"
  "errors"
  "strconv"
)

func main() {
  var inputFile string
  flag.StringVar(&inputFile, "i", "", "Specify the input filename")
  flag.StringVar(&inputFile, "input", "", "Specify the input filename")

  flag.Parse()

  if inputFile == "" {
    fmt.Println("Usage: go run part1.go -i/--input <filename>")
    return
  }

  file, err := os.Open(inputFile)
  if err != nil {
    fmt.Println("Error opening file:", err)
    return
  }
  defer file.Close()

  var lines []string

  scanner := bufio.NewScanner(file)
  for scanner.Scan() {
    line := scanner.Text()
    lines = append(lines, line)
  }

  solve(lines)

  if err := scanner.Err(); err != nil {
    fmt.Println("Error reading file:", err)
    return
  }
}

func solve(lines []string) error {
  var total = 0;
  for _, l := range lines {
    var startDigit int
    var endDigit int

    // get first digit
    for i := 0; i < len(l); i++ {
      if '0' <= l[i] && l[i] <= '9' {
        startDigit, _ = strconv.Atoi(l[i:i+1])
        break
      }
    }
    if startDigit == 0 {
      errorMessage := fmt.Sprintf("cannot find a digit from line: %s", l)
      return errors.New(errorMessage)
    }
    for i := len(l) - 1; i >= 0; i-- {
      if '0' <= l[i] && l[i] <= '9' {
        endDigit, _ = strconv.Atoi(l[i:i+1])
        break
      }
    }

    total += startDigit * 10 + endDigit
  }
  fmt.Println(total)
  return nil
}


